app/src/main/java/com/discordtime/gpts/maps/viewmodel/LocationManager.kt

```diff
@@ -0,0 +1,47 @@
+package com.discordtime.gpts.maps.viewmodel
+
+import android.content.Context
+import android.location.Location
+import android.os.Looper
+import com.google.android.gms.location.*
+
+class LocationManager(val mContext: Context, private val locationListener: LocationListener) {
+
+    private val UPDATE_INTERVAL: Long = 2 * 60 * 1000 //2 mins
+    private val FASTEST_INTERVAL: Long = 40 * 1000 //40 segs
+
+    private val mLocationRequest = LocationRequest()
+    private lateinit var mFusedLocationProviderClient: FusedLocationProviderClient
+
+    fun startLocationUpdates() {
+        //updates properties
+        mLocationRequest.priority = LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY
+        mLocationRequest.interval = UPDATE_INTERVAL
+        mLocationRequest.fastestInterval = FASTEST_INTERVAL
+
+        var builder = LocationSettingsRequest.Builder()
+        builder.addLocationRequest(mLocationRequest)
+        var locationSettingsRequest = builder.build()
+
+        //checks if location settings are satisfied
+        var settingsClient = LocationServices.getSettingsClient(mContext)
+        settingsClient.checkLocationSettings(locationSettingsRequest)
+
+        mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(mContext)
+        mFusedLocationProviderClient.requestLocationUpdates(mLocationRequest, object : LocationCallback() {
+            override fun onLocationResult(locationResult: LocationResult?) {
+                onLocationChanged()
+            }
+
+        }, Looper.myLooper())
+    }
+
+    private fun onLocationChanged() {
+        //var msg = "location changed: " + location?.latitude + ", " + location?.longitude
+        //Toast.makeText(mContext, msg, Toast.LENGTH_SHORT).show()
+
+        mFusedLocationProviderClient.lastLocation.addOnSuccessListener { location: Location ->
+            locationListener.onLocationChanged(location)
+        }
+    }
```

Comment:
This part is confusing. You're using the callback from the requestLocationUpdates to call the onLocationChanged function that defines another callback to call your callback. I don't really know now how to work with LocationServices so I don't know if you need both of "requestLocationUpdates" and "lasLocation.addOnSuccessListener" and I think you shouldn't define a listener inside the function that is called when the onLocationResult is called.

Make sure to read the documentation to have a better understanding of what each of these things do.
https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient.html

Reply:
Vinicius is partially right in my humble opinion. Allow me to explain:

`requestLocationUpdates` is for requesting further location updates, as configured by `locationRequest` object. As you coded it, location updates gotten from `mFusedLocationProviderClient.requestLocationUpdates` go to a `locationCallback`, and then you send to your `onLocationChanged`, which is **correct**

What's wrong is calling `mFusedLocationProviderClient.lastLocation` inside the `onLocationChanged` callback: a means to get the last known location (almost) immediately if it's available. Then you can save the location and do whatever you want on `onLocationChanged`

So this is the architecture:

```
mFusedLocationProviderClient.requestLocationUpdates _______.
mFusedLocationProviderClient.lastLocation _________________|____>    onLocationChanged 
```

To ways of getting the location, one single-shot, one recurring, both processed by `onLocationChanged`, where you should do what you want with the gotten location.

Reply:
e.g.
```
        mFusedLocationProviderClient.lastLocation.addOnSuccessListener { location: Location ->
		onLocationChanged(location)
        }

        mLocationRequest.interval = UPDATE_INTERVAL
        mLocationRequest.fastestInterval = FASTEST_INTERVAL

        var builder = LocationSettingsRequest.Builder()
        builder.addLocationRequest(mLocationRequest)
        var locationSettingsRequest = builder.build()

        //checks if location settings are satisfied
        var settingsClient = LocationServices.getSettingsClient(mContext)
        settingsClient.checkLocationSettings(locationSettingsRequest)

        mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(mContext)
        mFusedLocationProviderClient.requestLocationUpdates(mLocationRequest, object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult?) {
                onLocationChanged(locationResult.location)
            }

        }, Looper.myLooper())
    }

    private fun onLocationChanged(l: Location) {
        locationListener.onLocationChanged(location)
    }
```